src/topology/mod.rs

```diff
@@ -664,6 +670,18 @@ fn handle_errors(
         })
 }
 
+/// If the closure returns false, then the element is removed
+fn retain<T>(vec: &mut Vec<T>, mut retain_filter: impl FnMut(&mut T) -> bool) {
+    let mut i = 0;
+    while let Some(data) = vec.get_mut(i) {
+        if retain_filter(data) {
+            i += 1;
+        } else {
+            let _ = vec.remove(i);
+        }
+    }
+}
```

Comment:
This looks just like `Vec::retain` to me. Am I missing something or can we just use the standard library version?

Reply:
We need mutable access to the items, but `Vec::retain` gives us immutable. We could use `Vec::drain_filter` but it's a nightly feature at the moment.